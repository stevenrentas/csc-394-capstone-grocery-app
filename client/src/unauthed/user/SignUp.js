import React, {useState} from "react";
import "../../styles/style.css"

const SignUp = () => {
  const [credentials, setCredentials] = useState({
    emailAddress: '',
    password: '',
    confirmPassword: '',
    firstname: '',
    lastname:'',
    username:''
  });

  const [error, setError] = useState({
    emailAddress: '',
    confirmPassword: '',
    username:''
});

const onInputChange = e => {
  const { name, value } = e.target;
  setCredentials({
    ...credentials,
    [name]: value
  });
}

const validateCredentials = () => {
  if (credentials.password !== credentials.confirmPassword){
    setError({confirmPassword: "Passwords do not match."});
    return false;
  }

  else if (!/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(credentials.emailAddress)){
    setError({emailAddress: "Enter a valid email address."});
    return false;
  }

  setError({confirmPassword: ""});
  setError({emailAddress: ""});

  return true;
}
 
async function createAccount () {
  if (validateCredentials()){
    const db_columns = {
      "username": credentials.username,
      "first_name": credentials.firstname,
      "last_name": credentials.lastname,
      "email": credentials.emailAddress,
      "pword": credentials.password
    };
    const response = await fetch("/api/v1/adduser", {method: "POST", headers: {'Content-Type':'application/json'}, body:JSON.stringify(db_columns)})
    .catch(error => console.error(error));

    if (response.statusText === "OK"){
      console.log("succesfully created an account!");
    }
  }
}


    return (
      <div id="login">
        <input placeholder="First Name" id="credBox" name="firstname" onChange={onInputChange}></input>
        <input placeholder="Last Name" id="credBox" name="lastname" onChange={onInputChange}></input>
        <input placeholder="Username" id="credBox" name="username" onChange={onInputChange}></input>
        <input placeholder="Email Address" id="credBox" name="emailAddress" onChange={onInputChange}></input>
        {error.emailAddress && <p className='err'>{error.emailAddress}</p>}
        <input placeholder="Password" id="credBox" type="password" name="password" onChange={onInputChange}></input>
        <input placeholder="Verify Password" id="credBox" type="password" name="confirmPassword" onChange={onInputChange}></input>
        {error.confirmPassword && <p className='err'>{error.confirmPassword}</p>}
        <button id="authenticate" onClick={createAccount}>CREATE ACCOUNT</button>
        <p id="otherAuthPage"><a id="otherAuthPage" href={"/login"}>Already a member?</a></p>
      </div>
    );
  };
  
  export default SignUp;
  