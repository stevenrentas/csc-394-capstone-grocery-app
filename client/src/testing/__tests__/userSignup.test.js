import React from 'react';
import {render, fireEvent, screen, getByPlaceholderText} from '@testing-library/react';
import SignUp from '../../unauthed/user/SignUp';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

describe('User Signup Components <Signup.js>', () => {
    
   
    test('renders components and makes snapshot', () =>{
        const {container} = render(<SignUp/>);
        expect(container).toMatchSnapshot();
    });

    test('Should render input box for "First name"', () => {
        const { container } = render(<SignUp/>);
        const firstNameInput = container.querySelector('input[name="firstname"]');
        expect(firstNameInput).toBeInDocument;
    });

    test('Should render input box for "Last name"', () => {
        const { container } = render(<SignUp/>);
        const lastNameInput = container.querySelector('input[name="lastname"]');
        expect(lastNameInput).toBeInDocument;
    });

    test('Should render input box for "username"', () => {
        const { container } = render(<SignUp/>);
        const usernameInput = container.querySelector('input[name="username"]');
        expect(usernameInput).toBeInDocument;
    });
    
    test('Should render input box for "email"', () => {
        const { container } = render(<SignUp/>);
        const emailInput = container.querySelector('input[name="emailAddress"]');
        expect(emailInput).toBeInDocument;
    });

    test('Should render input box for "password"', () => {
        const { container } = render(<SignUp/>);
        const passwordInput = container.querySelector('input[name="password"]');
        expect(passwordInput).toBeInDocument;
    });

    test('Should render input box for "varify password"', () => {
        const { container } = render(<SignUp/>);
        const confirmPassInput = container.querySelector('input[name="confirmPassword"]');
        expect(confirmPassInput).toBeInDocument;
    });

    test('input boxes should be empty', ()=> {
        render(<SignUp/>);
        const firstnameInput = screen.getByPlaceholderText("First Name")
        const lastnameInput = screen.getByPlaceholderText("Last Name");
        const usernameInput = screen.getByPlaceholderText("Username");
        const emailInput = screen.getByPlaceholderText("Email Address");
        const passwordInput = screen.getByPlaceholderText("Password");
        const verifyInput = screen.getByPlaceholderText("Verify Password");

        expect(firstnameInput.value).toBe("");
        expect(lastnameInput.value).toBe("");
        expect(usernameInput.value).toBe("");
        expect(emailInput.value).toBe("");
        expect(passwordInput.value).toBe("");
        expect(verifyInput.value).toBe("");

    });

    test ('Should render a "Create account" button',() =>{
        const {getByText} = render(<SignUp/>);
        const createAccountButton = getByText("CREATE ACCOUNT");
        expect(createAccountButton).toBeInDocument;

    });

    test('onClick function called for "CREATE ACCOUNT" buton', () => {
        const onClickMock = jest.fn();
        const {getByText} = render( 
            <button onClick={onClickMock}>CREATE ACCOUNT</button>
        );
        const button = getByText('CREATE ACCOUNT');
        fireEvent.click(button);
        expect(onClickMock).toHaveBeenCalled();
    });

    test ('Should render a "Already a member?" button',() =>{
        const {getByText} = render(<SignUp/>);
        const alreadyMemButton = getByText("Already a member?");
        expect(alreadyMemButton).toBeInDocument;

    });

});
