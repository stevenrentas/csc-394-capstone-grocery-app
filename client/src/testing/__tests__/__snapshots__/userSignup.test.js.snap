// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Signup Components <Signup.js> renders components and makes snapshot 1`] = `
<div>
  <div
    id="login"
  >
    <input
      id="credBox"
      name="firstname"
      placeholder="First Name"
    />
    <input
      id="credBox"
      name="lastname"
      placeholder="Last Name"
    />
    <input
      id="credBox"
      name="username"
      placeholder="Username"
    />
    <input
      id="credBox"
      name="emailAddress"
      placeholder="Email Address"
    />
    <input
      id="credBox"
      name="password"
      placeholder="Password"
      type="password"
    />
    <input
      id="credBox"
      name="confirmPassword"
      placeholder="Verify Password"
      type="password"
    />
    <button
      id="authenticate"
    >
      CREATE ACCOUNT
    </button>
    <p
      id="otherAuthPage"
    >
      <a
        id="signupLink"
      >
        Already a member?
      </a>
    </p>
  </div>
</div>
`;
